### ISSUE TITLE
Add default keybindings for ctrl+backspace/ctrl+del for xterm-like terminals

### ISSUE THREAD

**Issue by ariasuni** [fish-shell/fish-shell#4770]
Original title of the bug report: «Can’t use begin/end/ctrl+backspace/ctrl+del with Konsole»

I use fish 2.7.1.

Konsole describe the keycode it sends here in a [configuration file](https://cgit.kde.org/konsole.git/tree/data/keyboard-layouts/default.keytab?id=5345bbb0afa8edd6ba4909eca2495745df36f6a1).

Here are the output I get from keystrokes:

– ctrl+del: `[3;5~`
– begin: `[1;5H`
– end: `[1;5F`

I don’t understand how to bind those keys manually, except for ctrl+backspace:
`bind \cH backward-kill-word`

**Comment by krader1961** [comment]
The easiest is to run `fish_key_reader` then press the key. It will show you a template. For example, when I press [ctrl-backspace]:

```
$ fish_key_reader                                               0 ms
Press a key

              hex:   7F  char: \x7F  (aka "del")
bind \x7F 'do something'
```

Put the example `bind` command in your `fish_user_key_bindings` function. Obviously substituting `do something` with an actual function documented in the `bind` man page or fish script. You can see existing bindings by just typing `bind`.

**Comment by ariasuni** [comment]
Thanks a lot!! Here’s my now working configuration:

```
bind \e\[3\;5~ kill-word
bind \cH backward-kill-word
bind \cV  beginning-of-line
bind \f end-of-line
```

**Comment by ariasuni** [comment]
I’m confused; fish aims at an out-of-the-box experience. So – correct me if I’m wrong – fish should include Konsole keybindings, if such presets don’t conflict with existing one.

I think that would be a nice improvements for users of Konsole to be able to use fish without spending a few hours to try to understand the mess that is terminal keybindings.

**Comment by faho** [comment]
>So – correct me if I’m wrong – fish should include Konsole keybindings

Sooo... I use Konsole as well, and I can't figure out which keys you mean here. I don't have a "begin" key - the (german) label says "Pos 1", and IIUC in english it's commonly known as the "home" key. That one sends `\e\[H` for me, and it's bound to "beginning-of-line" already - by default.

I have an "End" key (well, it says "Ende" on it), which sends `\e\[F`. It's bound to "end-of-line", by default.

Both of these match what the konsole config says, as long as you've set the "Default (XFree 4)" preset. If you change that, we have no way of figuring out which is which.

I think what you call "suppr" is called "page up" in english (or "pgup" for short). That by itself is bound for me, but ctrl-pgup isn't bound at all, for any terminal.

So what we need to figure out here is why these keys send different sequences for you - most likely you've selected a different preset in konsole's settings, which is something we pretty much can't support.

If you're asking for more friendly ways to bind these keys, there is #3018. The problem there though is that AFAIK, there is no central place or standard for what control+somekey should send, unless that "somekey" is a letter or number (in which case many of them don't do what you want). That's the unfortunate reality of how terminals work - it's 70s tech all the way down.

**Comment by ariasuni** [comment]
Sorry, it seems I got confused at some point. Begin and end keys works, but not Ctrl+Suppr and ([incoming](https://bugs.kde.org/show_bug.cgi?id=391075)) Ctrl+Backspace. I’m not sure what’s the policy for keybindings inclusion, but here’s what I use:

```sh
bind \e\[3\;5~ kill-word
bind \b backward-kill-path-component
```

**Comment by faho** [comment]
>I’m not sure what’s the policy for keybindings inclusion

There's no codified policy, but I'd imagine we'd all agree on something like the following:

- We prefer to use terminfo to look these things up. However, since terminfo is quite incomplete:

- We use hard-coded sequences where they don't break other terminals

- The binding should make some amount of sense and be useful generally - otherwise you can just add it yourself.

The second one here is unfortunately what kills your `\b` suggestion. Unfortunately, `\b` is the "backspace" sequence. Because of some historical confusion (the most boring argument since the US government had a kerfuffle about paper sizes), pressing the backspace key mostly sends the "delete char" sequence. But sometimes it still sends actual backspace, and some people like to configure this, and some terminal authors like to use one over the other (for reasons I don't understand).  And it's not really possible to detect what exactly your terminal does.

So effectively, for us, "backspace" (== `\b`) and "delete char" (== `\x7f`) should be kept in sync. That means `\b` should do "backward-delete-char" like `\x7f` does.

For ctrl+pgup, that has no entry in the key database we use (terminfo, which is the standard for this sort of thing). However, konsole identifies as xterm (by setting $TERM to "xterm-something"), and xterm sends the same sequence.

I am not, however, sold on the particular binding - on my keyboard layout, pressing ctrl+pgup requires grabbing across about a third of the keyboard. Which do you use? Azerty or something? I'm using QWERTZ.

It might be possible to add that key->sequence mapping ourselves, and make "ctrl+pgup" available as an argument to `bind`, so you'd do `bind -k ctrl+pgup kill-word`. I'd track that in #3018 (because allowing something like that is a prerequisite to that).

TL;DR: Terminals suck, and there's no way around it.

**Comment by ariasuni** [comment]
Sorry, I was talking about ctrl+del. Anyway, I perfectly understand your point.

Konsole tries to work like xterm, and I saw that fish set different keybindings when terminal declares it’s xterm. Is it possible to put those keybindings here then, or am I missing something?

**Comment by faho** [comment]
>I saw that fish set different keybindings when terminal declares it’s xterm

Currently, not quite. The changes happening for xterm currently are all because of terminfo.

But it's certainly possible to hardcode this sequence if $TERM = xterm.

Ideally, someone would already have collated a database of all of these mappings for us to just include and be done.

>Sorry, I was talking about ctrl+del.

Aah, that makes much more sense.

**Comment by ariasuni** [comment]
So, if I understand correctly, the terminfo database should be updated to reflect more accurately the behavior of Konsole?

**Comment by faho** [comment]
Terminfo doesn't have a field for ctrl+pgup or ctrl+del, so those would have to be added. And in my experience terminfo is quite conservative, so this most likely won't happen.

**Comment by ariasuni** [comment]
Could we bind the sequences I mentioned only when a xterm-like terminal is used? (xterm, GNOME Terminal, Konsole, maybe others?)

Sorry if I sound insistent; I just don’t quite understand everything and I’d like to find a solution that satisfy everyone. :)

**Comment by faho** [comment]
>Could we bind the sequences I mentioned only when a xterm-like terminal is used? (xterm, GNOME Terminal, Konsole, maybe others?)

Yes, of course. We already do something like this for some rxvt-specific sequences. See https://github.com/fish-shell/fish-shell/blob/master/share/functions/fish_default_key_bindings.fish#L90-L97.

All that's needed code-wise is to add

```fish
    case 'xterm*'
        # ctrl-delete
        bind \e\[3\;5~ kill-word
```

That's what I meant by "But it's certainly possible to hardcode this sequence if $TERM = xterm.".

However, there's a few remaining questions:

We currently divide our bindings into "default", "vi" and "shared" (i.e. the stuff you get in both the emacs-ish "default" mode and in vi-mode). Which should this fall into? Is this a binding that's useful in vi-mode? (which is something we never really nailed down)

Should we instead do

```fish
set -l ctrl_del
switch "$TERM"
    case 'xterm*'
        set ctrl_del \e\[3\;5~
    case 'rxvt*'
        set ctrl_del \e\[3^

# and later

set -q ctrl_del[1]; and bind $ctrl_del kill-word
```

?

That's more extensible, but also bigger and more intrusive.

**Comment by ariasuni** [comment]
As far as I know, ctrl+del and ctrl+backspace don’t conflict with vi keybindings, so they should probably go in «shared».

**Comment by BarbzYHOOL** [comment]
I haven't read everything but I'm trying to set "ctrl+backspace" to delete whole word, but when itype "fish_key_reader" and I press ctrl+backspace, it is the same result as backspace, so it means I can not differentiate ctrl+backspace and that sux balls

How to proceed?

**Comment by laneme** [comment]
Dead thread? looking for it in 2024

**Comment by krobelus** [comment]
this is fixed in master
